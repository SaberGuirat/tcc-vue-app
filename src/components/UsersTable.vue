<template>
  <v-container>
    <v-card>
      <v-card-title>
        <v-text-field
          v-model="search"
          append-icon="mdi-magnify"
          label="Search"
          single-line
          hide-details
        ></v-text-field>
      </v-card-title>
      <v-data-table :headers="headers" :items="persons" :search="search">
        <template v-slot:top>
          <v-toolbar flat>
            <v-toolbar-title>Users Infos</v-toolbar-title>
            <v-divider class="mx-4" inset vertical></v-divider>
            <v-spacer></v-spacer>
            <v-dialog v-model="dialog" max-width="500px">
              <template v-slot:activator="{ on, attrs }">
                <v-btn
                  color="primary"
                  dark
                  class="mb-2"
                  v-bind="attrs"
                  v-on="on"
                >
                  Add User
                </v-btn>
              </template>
              <v-card>
                <v-card-title>
                  <span class="text-h5">{{ formTitle }}</span>
                </v-card-title>
                <v-card-text>
                  <v-container>
                    <v-row>
                      <v-col cols="12" sm="6" md="3">
                        <v-text-field
                          v-model="editedPerson.firstName"
                          label="First Name"
                        ></v-text-field>
                      </v-col>
                      <v-col cols="12" sm="6" md="3">
                        <v-text-field
                          v-model="editedPerson.lastName"
                          label="Last Name"
                        ></v-text-field>
                      </v-col>
                      <v-col cols="12" sm="12" md="6">
                        <v-text-field
                          v-model="editedPerson.email"
                          label="Email"
                        ></v-text-field>
                      </v-col>
                    </v-row>
                  </v-container>
                </v-card-text>
                <v-card-actions>
                  <v-spacer></v-spacer>
                  <v-btn color="blue darken-1" text @click="close">
                    Cancel
                  </v-btn>
                  <v-btn color="blue darken-1" text @click="save">
                    Save
                  </v-btn>
                </v-card-actions>
              </v-card>
            </v-dialog>
            <v-dialog v-model="dialogDelete" max-width="500px">
              <v-card>
                <v-card-title class="text-h5"
                  >Are you sure you want to delete this User?</v-card-title
                >
                <v-card-actions>
                  <v-spacer></v-spacer>
                  <v-btn color="blue darken-1" text @click="closeDelete"
                    >Cancel</v-btn
                  >
                  <v-btn color="blue darken-1" text @click="deletePersonConfirm"
                    >OK</v-btn
                  >
                  <v-spacer></v-spacer>
                </v-card-actions>
              </v-card>
            </v-dialog>
          </v-toolbar>
        </template>
        <template v-slot:[`item.actions`]="{ item }">
          <v-icon small class="mr-2" @click="editPerson(item)">
            mdi-pencil
          </v-icon>
          <v-icon small @click="deletePerson(item)">
            mdi-delete
          </v-icon>
        </template>
        <template v-slot:no-data>
          <v-btn color="primary" @click="initialize">
            Reset
          </v-btn>
        </template>
      </v-data-table>
    </v-card>
  </v-container>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";
import { IPerson } from "../interfaces/person.interface";

@Component({
  props: {
    persons: Array as () => IPerson[],
    headers: Array,
    addUser: Function,
    deleteUser: Function,
    updateUser: Function,
  },
  watch: {
    dialog(val) {
      val || this.close();
    },
    dialogDelete(val) {
      val || this.closeDelete();
    },
  },
})
export default class UsersTable extends Vue {
  persons: IPerson[];
  editedIndex = -1;
  editedPerson: IPerson = {
    firstName: "",
    lastName: "",
    email: "",
  };
  defaultPerson: IPerson = {
    firstName: "",
    lastName: "",
    email: "",
  };
  dialog = false;
  dialogDelete = false;

  data() {
    return {
      search: "",
    };
  }

  get formTitle() {
    return this.editedIndex === -1 ? "Add User" : "Edit User";
  }

  editPerson(person: IPerson) {
    this.editedIndex = this.persons.indexOf(person);
    this.editedPerson = Object.assign({}, person);
    this.dialog = true;
  }

  deletePerson(person: IPerson) {
    this.editedIndex = this.persons.indexOf(person);
    this.editedPerson = Object.assign({}, person);
    this.dialogDelete = true;
  }

  deletePersonConfirm() {
    this.$props.deleteUser(this.persons[this.editedIndex]._id);
    this.closeDelete();
  }

  close() {
    this.dialog = false;
    this.$nextTick(() => {
      this.editedPerson = Object.assign({}, this.defaultPerson);
      this.editedIndex = -1;
    });
  }

  closeDelete() {
    this.dialogDelete = false;
    this.$nextTick(() => {
      this.editedPerson = Object.assign({}, this.defaultPerson);
      this.editedIndex = -1;
    });
  }

  save() {
    if (this.editedIndex > -1) {
      this.$props.updateUser(
        this.persons[this.editedIndex]._id,
        this.editedPerson
      );
    } else {
      this.$props.addUser(this.editedPerson);
    }
    this.close();
  }

  initialize() {
    this.persons = [];
  }
}
</script>
