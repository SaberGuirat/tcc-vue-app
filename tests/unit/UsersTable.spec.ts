import { mount, createLocalVue } from "@vue/test-utils";
import UsersTable from "@/components/UsersTable.vue";
import Vue from "vue";
import Vuetify from "vuetify";
import { IPerson } from "@/interfaces/person.interface";

const persons: IPerson[] = [
  {
    _id: "123",
    firstName: "saber",
    lastName: "Guirat",
    email: "saber.guirat92@gmail.com",
  },
];

const headers = [
  {
    text: "ID",
    align: "start",
    filterable: true,
    value: "_id",
  },
  { text: "First Name", value: "firstName" },
  { text: "Last Name", value: "lastName" },
  { text: "Email", value: "email" },
  { text: "Actions", value: "actions", sortable: false },
];
const addUser = (person: IPerson) => {
  console.log("add user");
};
const deleteUser = () => {
  console.log("delete user");
};
const updateUser = () => {
  console.log("update user");
};

describe("UsersTable.vue", () => {
  const localVue = createLocalVue();
  let vuetify: any;
  Vue.use(Vuetify);

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const mountFunction = (options: any) => {
    return mount(UsersTable, {
      localVue,
      vuetify,
      ...options,
    });
  };

  it("renders props when passed", async () => {
    const wrapper = mountFunction({
      propsData: { headers, persons, addUser, deleteUser, updateUser },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
