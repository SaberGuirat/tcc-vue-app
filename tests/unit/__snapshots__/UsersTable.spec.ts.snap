// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTable.vue renders props when passed 1`] = `
<div class="container">
  <div class="v-card v-sheet theme--light">
    <div class="v-card__title">
      <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-3" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-3" type="text"></div>
            <div class="v-input__append-inner">
              <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-magnify theme--light"></i></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-data-table v-data-table--has-top v-data-table--has-bottom theme--light">
      <header class="v-sheet theme--light v-toolbar v-toolbar--flat" style="height: 64px;">
        <div class="v-toolbar__content" style="height: 64px;">
          <div class="v-toolbar__title">Users Infos</div>
          <hr role="separator" aria-orientation="vertical" class="mx-4 v-divider v-divider--inset v-divider--vertical theme--light">
          <div class="spacer"></div>
          <div role="dialog" class="v-dialog__container"><button type="button" class="mb-2 v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--default primary" role="button" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content">
                Add User
              </span></button>
            <!---->
          </div>
          <div role="dialog" class="v-dialog__container">
            <!---->
          </div>
        </div>
      </header>
      <div class="v-data-table__wrapper">
        <table>
          <colgroup>
            <col class="">
            <col class="">
            <col class="">
            <col class="">
            <col class="">
          </colgroup>
          <thead class="v-data-table-header">
            <tr>
              <th role="columnheader" scope="col" aria-label="ID: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>ID</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="First Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>First Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Last Name: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Last Name</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Email: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Email</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
              <th role="columnheader" scope="col" aria-label="Actions" class="text-start"><span>Actions</span></th>
            </tr>
          </thead>
          <tbody>
            <tr class="">
              <td class="text-start">123</td>
              <td class="text-start">saber</td>
              <td class="text-start">Guirat</td>
              <td class="text-start">saber.guirat92@gmail.com</td>
              <td class="text-start"><button type="button" class="v-icon notranslate mr-2 v-icon--link mdi mdi-pencil theme--light" style="font-size: 16px;"></button> <button type="button" class="v-icon notranslate v-icon--link mdi mdi-delete theme--light" style="font-size: 16px;"></button></td>
            </tr>
          </tbody>
        </table>
      </div>
      <div class="v-data-footer">
        <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-21" class="v-input__slot">
                <div class="v-select__slot">
                  <div class="v-select__selections">
                    <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-21" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                  </div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden" value="10">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="v-data-footer__pagination">1-1 of 1</div>
        <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
        <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
      </div>
    </div>
  </div>
</div>
`;
